index.js

```diff
@@ -224,44 +186,103 @@ module.exports = function (config) {
+    if (typeof plugin === 'undefined') return koop.log.error('Plugin undefined, skipping registration.')
+    if (!plugin.name) return koop.log.error('Plugin missing name, skipping registration.')
 
-  koop.Cache = new koop.DataCache(koop)
+    if (plugin.type) {
+      if (plugin.type === 'provider') return app.registerProvider(plugin)
+      if (plugin.type === 'cache') return app.registerCache(plugin)
+      if (plugin.type === 'plugin') return app.registerPlugin(plugin)
 
-  // use the default local cache until a DB adapter mod is registered
-  if (!config.db || !config.db.conn) {
-    console.warn('Warning koop w/o persistent cache means no data will be cached across server sessions.')
+      koop.log.warn('Unrecognized plugin type. Defaulting to provider.')
+      return app.registerProvider(plugin)
+    }
+
+    koop.log.warn('Plugin missing type property. Defaulting to provider.')
+    app.registerProvider(plugin)
   }
 
-  // the default cache is the local in-mem cache
-  // to persist data you must call registerCache with db adapter
-  koop.Cache.db = koop.LocalDB
+  /**
+   * registers koop providers
+   * exposes the provider's routes, controller, and model
+   * @param {object} provider - the provider to be registered
+   */
+  app.registerProvider = function (provider) {
+    app.services[provider.name] = provider
+
+    var model = new provider.model(koop) // eslint-disable-line
+    var controller = new provider.controller(model, koop.BaseController) // eslint-disable-line
+
+    // if a provider has a status object store it
+    if (provider.status) {
+      app.status.providers[provider.name] = provider.status
+    }
+
+    // binds a series of standard routes
+    if (provider.name && provider.pattern) {
+      app._bindDefaultRoutes(provider.name, provider.pattern, controller)
+    }
+
+    // add each route, the routes let us override defaults etc.
+    app._bindRoutes(provider.routes, controller)
+  }
 
   /**
-   * Register DB adapters into the main Koop app
+   * registers a koop cache
    * overwrites any existing koop.Cache.db
-   * @param {object} a koop db adapter
+   * @param {object} cache - a koop database adapter
    */
-  app.registerCache = function (adapter) {
-    if (config.db && config.db.conn) {
-      koop.Cache.db = adapter.connect(config.db.conn, koop)
-    } else {
-      console.log('Cannot register this cache, missing db connection in config')
+  app.registerCache = function (cache) {
+    if (!config.db || !config.db.conn) {
+      return koop.log.error('Cannot register cache. Missing db.conn property in config.')
     }
-    return
+
+    koop.Cache.db = cache.connect(config.db.conn, koop)
   }
 
-  return app
+  /**
+   * registers a koop plugin
+   * Plugins can be any function that you want to have global access to
+   * within koop provider models
+   * @param {object} any koop plugin
+   */
+  app.registerPlugin = function (plugin) {
+    koop[plugin.name] = plugin
+  }
```

Comment:
This method needs to be fleshed out if we want proper plugin support. I exposed `defaultRoutes` on the koop object above as a way of potentially letting plugins add routes to all providers for tiles, preview, etc.